salt/modules/virt.py

```diff
@@ -599,18 +601,29 @@ def _gen_xml(
                 context["boot"]["kernel"] = "/usr/lib/grub2/x86_64-xen/grub.xen"
                 context["boot_dev"] = []
 
-    if "serial_type" in kwargs:
-        context["serial_type"] = kwargs["serial_type"]
-    if "serial_type" in context and context["serial_type"] == "tcp":
-        if "telnet_port" in kwargs:
-            context["telnet_port"] = kwargs["telnet_port"]
-        else:
-            context["telnet_port"] = 23023  # FIXME: use random unused port
-    if "serial_type" in context:
-        if "console" in kwargs:
-            context["console"] = kwargs["console"]
-        else:
-            context["console"] = True
+    context["serials"] = []
+    for serial in enumerate(serials):
+        serial_context = {
+            "type": serial.type
+        }
+        if serial_context["type"] == "tcp":
+            if (serial["port"] is None) or (serial["protocol"] is None):
+                raise CommandExecutionError("Absent serial parameter for TCP connection")
+            serial_context["port"] = serial["port"]
+            serial_context["protocol"] = serial["protocol"]
```

Comment:
Use `get` since it provides a default value even if the property you are looking for is not set.

And I don't agree with this check. The old code was using a default port `23023` if not set and forced the protocol to `telnet` in the Jinja file. Even if that is not perfect you should provide the same feature.

``` suggestion
            serial_context["port"] = serial.get("port", default_port)
            serial_context["protocol"] = serial.get("protocol", "telnet")
```